File: backend/src/ai/ai.service.ts
import { Injectable, InternalServerErrorException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { GoogleGenerativeAI, GenerativeModel } from '@google/generative-ai';
import { DebugCodeDto } from './dto/debug-code.dto';
import { ExplainCodeDto } from './dto/explain-code.dto';

@Injectable()
export class AiService {
  private genAI: GoogleGenerativeAI | null = null;
  private readonly model: string;

  constructor(private readonly config: ConfigService) {
    this.model = this.config.get<string>('GEMINI_MODEL') || 'gemini-2.5-flash';
  }

  private getModel(): GenerativeModel {
    const apiKey = this.config.get<string>('GEMINI_API_KEY');
    if (!apiKey) {
      throw new InternalServerErrorException(
        'GEMINI_API_KEY is not set. Add it to your .env file to use AI features.',
      );
    }
    if (!this.genAI) {
      this.genAI = new GoogleGenerativeAI(apiKey);
    }
    return this.genAI.getGenerativeModel({ model: this.model });
  }

  async debugCode(dto: DebugCodeDto) {
    const model = this.getModel();

    const prompt = [
      `You are an expert ${dto.language || ''} debugger helping a developer inside "DevAgent", a developer workspace tool.`,
      `Analyze the following code${dto.errorMessage ? ' and the error message below' : ''} and identify the bug(s).`,
      '',
      '```' + (dto.language || ''),
      dto.code,
      '```',
      dto.errorMessage ? `\nError message:\n${dto.errorMessage}` : '',
      dto.context ? `\nAdditional context:\n${dto.context}` : '',
      '',
      'Respond in this exact structure:',
      '1. **Root cause** — a concise explanation of what is wrong.',
      '2. **Fix** — the corrected code in a fenced code block.',
      '3. **Why it works** — a short explanation of the fix.',
      '4. **Prevention tip** — one tip to avoid this class of bug in the future.',
    ]
      .filter(Boolean)
      .join('\n');

    const result = await model.generateContent(prompt);
    return { result: result.response.text() };
  }

  async explainCode(dto: ExplainCodeDto) {
    const model = this.getModel();
    const level = dto.level || 'intermediate';

    const prompt = [
      `Explain the following ${dto.language || ''} code to a developer at a ${level} level.`,
      'Cover: what it does overall, how it works step by step, and call out any notable',
      'patterns, edge cases, or potential issues.',
      '',
      '```' + (dto.language || ''),
      dto.code,
      '```',
    ].join('\n');

    const result = await model.generateContent(prompt);
    return { result: result.response.text() };
  }

  async analyzeResume(base64DataUrl: string): Promise<string> {
    const model = this.getModel();

    const matches = base64DataUrl.match(/^data:(.+);base64,(.+)$/);
    if (!matches) {
      throw new Error('Invalid data URL format. Expected base64 encoded document.');
    }
    const mimeType = matches[1];
    const base64Data = matches[2];

    const prompt = `
Analyze the attached resume and extract the key details into a structured JSON format.
Ensure the output is valid JSON (and only JSON, no markdown formatting blocks, no backticks).
Structure:
{
  "name": "Full Name",
  "profession": "Job Title (e.g. Full Stack Developer)",
  "contact": {
    "email": "Email address",
    "phone": "Phone number",
    "location": "Location (e.g. Salem, Tamil Nadu)",
    "github": "GitHub profile link",
    "linkedin": "LinkedIn profile link",
    "portfolio": "Portfolio link"
  },
  "summary": "Concise professional summary...",
  "skills": ["Skill 1", "Skill 2", ...],
  "experience": [
    {
      "role": "Job Title",
      "company": "Company Name",
      "duration": "Duration (e.g. 2024)",
      "description": "Short description of responsibilities and achievements"
    }
  ],
  "projects": [
    {
      "title": "Project Title",
      "description": "Short description of what the project does",
      "tech": ["Next.js", "React", ...]
    }
  ],
  "education": [
    {
      "degree": "Degree (e.g., B.Tech. in Information Technology)",
      "school": "University Name",
      "duration": "Duration (e.g., 2023 - 2027)",
      "score": "Score (e.g. CGPA: 8.1 or 73%)"
    }
  ],
  "certifications": [
    {
      "name": "Certification Name",
      "authority": "Issuing Authority (e.g. NVIDIA)",
      "date": "Date"
    }
  ]
}
`;

    const docPart = {
      inlineData: {
        data: base64Data,
        mimeType: mimeType,
      },
    };

    const result = await model.generateContent([prompt, docPart]);
    const text = result.response.text();
    const cleanText = text.replace(/^```json\s*/, '').replace(/```\s*$/, '').trim();
    return cleanText;
  }
}

File: backend/src/main/java/com/devagent/ai/AiService.java
package com.devagent.ai;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import okhttp3.*;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.io.IOException;

@Service
public class AiService {

    private static final String BASE =
            "https://generativelanguage.googleapis.com/v1beta/models/";
    private static final MediaType JSON = MediaType.get("application/json");

    private final OkHttpClient httpClient = new OkHttpClient.Builder()
            .callTimeout(java.time.Duration.ofSeconds(120))
            .build();
    private final ObjectMapper mapper = new ObjectMapper();

    @Value("${gemini.api-key:}")
    private String apiKey;

    @Value("${gemini.model:gemini-2.5-flash}")
    private String model;

    // ── Core call ────────────────────────────────────────────────────────────

    private String generate(ObjectNode requestBody) {
        if (apiKey == null || apiKey.isBlank()) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
                    "GEMINI_API_KEY is not set.");
        }
        String url = BASE + model + ":generateContent?key=" + apiKey;
        try {
            RequestBody body = RequestBody.create(mapper.writeValueAsString(requestBody), JSON);
            Request req = new Request.Builder().url(url).post(body).build();
            try (Response res = httpClient.newCall(req).execute()) {
                String responseStr = res.body().string();
                if (!res.isSuccessful()) {
                    throw new ResponseStatusException(HttpStatus.BAD_GATEWAY,
                            "Gemini API error: " + res.code() + " " + responseStr);
                }
                JsonNode root = mapper.readTree(responseStr);
                return root.path("candidates").get(0)
                        .path("content").path("parts").get(0)
                        .path("text").asText();
            }
        } catch (ResponseStatusException e) {
            throw e;
        } catch (IOException e) {
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY,
                    "Failed to reach Gemini API: " + e.getMessage());
        }
    }

    private ObjectNode textRequest(String prompt) {
        ObjectNode body = mapper.createObjectNode();
        ArrayNode contents = body.putArray("contents");
        ObjectNode content = contents.addObject();
        ArrayNode parts = content.putArray("parts");
        parts.addObject().put("text", prompt);
        ObjectNode config = body.putObject("generationConfig");
        config.put("maxOutputTokens", 2000);
        return body;
    }

    // ── Public methods ────────────────────────────────────────────────────────

    public String debugCode(String code, String language, String errorMessage, String context) {
        String lang = language != null ? language : "";
        String prompt = String.join("\n",
                "You are an expert " + lang + " debugger helping a developer inside \"DevAgent\".",
                "Analyze the following code" + (errorMessage != null ? " and error below" : "") + " and identify bugs.",
                "",
                "```" + lang,
                code,
                "```",
                errorMessage != null ? "\nError message:\n" + errorMessage : "",
                context != null ? "\nAdditional context:\n" + context : "",
                "",
                "Respond in this exact structure:",
                "1. **Root cause** — concise explanation of what is wrong.",
                "2. **Fix** — corrected code in a fenced code block.",
                "3. **Why it works** — short explanation of the fix.",
                "4. **Prevention tip** — one tip to avoid this class of bug."
        );
        return generate(textRequest(prompt));
    }

    public String explainCode(String code, String language, String level) {
        String lang = language != null ? language : "";
        String lvl = level != null ? level : "intermediate";
        String prompt = String.join("\n",
                "Explain the following " + lang + " code to a developer at a " + lvl + " level.",
                "Cover: what it does overall, how it works step by step, and notable patterns, edge cases, or issues.",
                "",
                "```" + lang,
                code,
                "```"
        );
        return generate(textRequest(prompt));
    }

    public String analyzeResume(String base64DataUrl) {
        if (!base64DataUrl.contains(";base64,")) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                    "Invalid data URL format.");
        }
        String[] parts = base64DataUrl.split(";base64,", 2);
        String mimeType = parts[0].replace("data:", "");
        String base64Data = parts[1];

        String promptText = """
                Analyze the attached resume and extract key details into valid JSON only (no markdown).
                Structure: { "name", "profession", "contact": { "email","phone","location","github","linkedin","portfolio" },
                "summary", "skills": [], "experience": [{"role","company","duration","description"}],
                "projects": [{"title","description","tech":[]}],
                "education": [{"degree","school","duration","score"}],
                "certifications": [{"name","authority","date"}] }
                """;

        ObjectNode body = mapper.createObjectNode();
        ArrayNode contents = body.putArray("contents");
        ObjectNode content = contents.addObject();
        ArrayNode partsList = content.putArray("parts");

        partsList.addObject().put("text", promptText);

        ObjectNode filePart = partsList.addObject();
        ObjectNode inlineData = filePart.putObject("inlineData");
        inlineData.put("mimeType", mimeType);
        inlineData.put("data", base64Data);

        return generate(body);
    }
}
